packages/worker/src/workflow/reusable-vm.ts

```diff
@@ -0,0 +1,250 @@
+    const { _context } = this;
+    if (_context == null) {
+      throw new IllegalStateError('Tried to use v8 context after Workflow creator was destroyed');
+    }
+    return _context;
+  }
+  /**
+   * Create a workflow with given options
+   */
+  async createWorkflow(options: WorkflowCreateOptions): Promise<Workflow> {
+    const context = this.context;
+    const bag: Record<string, unknown> = {};
+    const activationContext = { isReplaying: options.info.unsafe.isReplaying };
+    const { isolateExecutionTimeoutMs, contextKeysToPreserve } = this;
+    const workflowModule: WorkflowModule = new Proxy(
+      {},
+      {
+        get(_: any, fn: string) {
+          return (...args: any[]) => {
+            Object.assign(context, bag);
+            // runInContext does not accept args, pass via globals
+            context.__TEMPORAL_ARGS__ = args;
+            try {
+              return vm.runInContext(`__TEMPORAL__.api.${fn}(...__TEMPORAL_ARGS__)`, context, {
+                timeout: isolateExecutionTimeoutMs,
+                displayErrors: true,
+              });
+            } finally {
+              const keysToDelete = [];
+              // TODO: non-enumerable global properties?
+              // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects
+              for (const k in context) {
+                if (!contextKeysToPreserve.has(k)) {
+                  bag[k] = context[k];
+                  context[k] = undefined;
+                  keysToDelete.push(k);
+                }
+              }
+              for (const k in keysToDelete) {
+                delete context[k];
+              }
+              // Need to preserve this for the unhandledRejection handler.
+              // TODO: There's probably a better way but this is simplest since we want to maintain compatibility with
+              // the non-reusable vm implementation.
+              context.__TEMPORAL_ACTIVATOR__ = bag.__TEMPORAL_ACTIVATOR__;
+            }
+          };
+        },
+      }
+    ) as any;
+
+    workflowModule.initRuntime({ ...options, sourceMap: this.workflowBundle.sourceMap });
+    const activator = bag.__TEMPORAL_ACTIVATOR__ as any;
+
+    const newVM = new ReusableVMWorkflow(
+      options.info,
+      context,
+      activator,
+      workflowModule,
+      isolateExecutionTimeoutMs,
+      this.hasSeparateMicrotaskQueue,
+      activationContext
+    );
+    ReusableVMWorkflowCreator.workflowByRunId.set(options.info.runId, newVM);
+    return newVM;
+  }
+
+  /**
+   * Inject console.log and friends into the Workflow isolate context.
+   *
+   * Overridable for test purposes.
+   */
+  protected injectConsole(): void {
+    this.context.console = {
+      log: (...args: any[]) => {
+        const { info } = this.context.__TEMPORAL_ACTIVATOR__;
+        if (info.isReplaying) return;
+        console.log(`[${info.workflowType}(${info.workflowId})]`, ...args);
+      },
```

Comment:
Could just wrap this in a little closure that takes the method to call on `console` since everything else is identical

Reply:
I'd move this to vm-shared.ts and use it for regular vm too. Then we can remove `ActivationContext` everywhere.

Reply:
Yeah.. I can dedupe this a bit.